feat(pii): Implement smart PII scrubbing for `logentry.formatted`

olksdr opened the pull request:
- Remove $logentry.formatted from `DATASCRUBBER_IGNORE` in `convert.rs`
- Implement `smart_scrub_logentry_formatted()` function that:
  - Replaces emails with `[email]`
  - Replaces credit cards with `[creditcard]`
  - Replaces SSNs with `[ssn]`
  - Replaces IBANs with `[iban]`
  - Applies user-configured PII rules

Avoid replacing the entire message and preserve as much context as possible

Fixes INGEST-464

olksdr committed:
Implement smart PII scrubbing for logentry.formatted

- Remove $logentry.formatted from DATASCRUBBER_IGNORE in convert.rs
- Implement smart_scrub_logentry_formatted() function that:
  - Replaces emails with [Email]
  - Replaces credit cards with [CreditCard]
  - Replaces SSNs with [SSN]
  - Replaces IBANs with [IBAN]
  - Applies user-configured PII rules
- Integrate smart scrubbing into PiiProcessor.process_string()
- Avoid replacing the entire message with [Filtered]

olksdr committed:
Add changelog entry

loewenheim commented on relay-pii/src/processor.rs:
```diff
@@ -16,9 +16,112 @@ use relay_protocol::{Annotated, Array, Meta, Remark, RemarkType, Value};
+    for pattern in &patterns {
+        let new_value = pattern.regex.replace_all(value, pattern.replacement);
+        if matches!(new_value, std::borrow::Cow::Owned(_)) {
+            *value = new_value.into_owned();
```
```suggestion
        if let std::borrow::Cow::Owned(new_value) = new_value {
            *value = new_value;
```

loewenheim commented on relay-pii/src/processor.rs:
```diff
@@ -16,9 +16,112 @@ use relay_protocol::{Annotated, Array, Meta, Remark, RemarkType, Value};
+    }
+
+    // Add all remarks for applied patterns and set original length if needed.
+    if !applied_patterns.is_empty() {
```
You don't need to check the emptiness of `applied_patterns` separately; if it is empty then nothing will happen in the `for suffix …` block or the `if original_length …` block.
```suggestion
```

loewenheim commented on relay-pii/src/processor.rs:
```diff
@@ -16,9 +16,112 @@ use relay_protocol::{Annotated, Array, Meta, Remark, RemarkType, Value};
+            ));
+        }
+
+        // Set original length if the length changed.
```
```suggestion
```

loewenheim commented on relay-pii/src/processor.rs:
```diff
@@ -16,9 +16,112 @@ use relay_protocol::{Annotated, Array, Meta, Remark, RemarkType, Value};
+    }
+
+    // Apply user-configured PII rules, but avoid rules that would completely filter logentry.formatted.
+    // We skip @anything rules and only apply pattern-based rules to preserve context.
```
This comment is redundant with the one on line 94.
```suggestion
```

loewenheim commented on relay-pii/src/processor.rs:
```diff
@@ -16,9 +16,112 @@ use relay_protocol::{Annotated, Array, Meta, Remark, RemarkType, Value};
+                if matches!(rule.ty, RuleType::RedactPair(_)) {
+                    if let Redaction::Replace(replace_redaction) = &rule.redaction {
+                        if replace_redaction.text == "[Filtered]" {
+                            continue;
```
Maybe nicer with a let chain instead of nested ifs.
```suggestion
                if matches!(rule.ty, RuleType::RedactPair(_))
                    && let Redaction::Replace(replace_redaction) = &rule.redaction
                    && replace_redaction.text == "[Filtered]"
                {
                    continue;
```

loewenheim commented on relay-pii/src/processor.rs:
```diff
@@ -1177,11 +1290,65 @@ mod tests {
+
+        // Should preserve context
+        assert!(formatted_str.contains("User"));
+        assert!(formatted_str.contains("failed login"));
```
I think it might be easier to tell what the end result should look like with a snapshot.

olksdr committed:
Add Bearer token scrubbing.

Also added another test to make sure that logentry.formatted is not scrubbed
even when we have a word "password" in there.

olksdr committed:
simplify after review

olksdr committed:
Added integration test

olksdr committed:
tests

Dav1dde commented on relay-pii/src/processor.rs:
```diff
@@ -153,6 +256,14 @@ impl Processor for PiiProcessor<'_> {
+            if path_str.eq_ignore_ascii_case("logentry.formatted") {
+                return smart_scrub_logentry_formatted(value, meta, state, self.compiled_config);
+            }
+        }
```
I don't think that's the way to do it, I think this should be possible to express by injecting a new default rule when converting from datascrubbing settings to the pii config which only matches on `logentry.formatted`.

olksdr replied:
We can quickly talk about this. 
I still wanna discuss some tradeoffs of this approach, and if we really wanna do something like this.

olksdr committed:
Use rulesets and specs to remove PII when scrubbing enabled

olksdr committed:
Rename the rule

olksdr committed:
Merge branch 'master' into fix/pii-logentry-formatted

loewenheim commented on CHANGELOG.md:
```diff
@@ -5,6 +5,7 @@
 **Features**:
 
 - Always emit a span usage metric, independent of span feature flags. ([#4976](https://github.com/getsentry/relay/pull/4976))
+- Support PII scrubbing in `logentry.formatted` avoiding replacing the entire value, still preserving the context. ([#4985](https://github.com/getsentry/relay/pull/4985))
```
```suggestion
- Avoid replacing the entire value of `logentry.formatted` during PII scrubbing. ([#4985](https://github.com/getsentry/relay/pull/4985))
```

olksdr committed:
Add proper builtin selector for bearer token

olksdr committed:
Add test to check if the common filters apply to the logentry.formatted

Dav1dde commented on relay-pii/src/regexes.rs:
```diff
@@ -60,8 +60,16 @@ pub fn get_regex_for_rule_type(
-            smallvec![(kv, &*PASSWORD_KEY_REGEX, ReplaceBehavior::replace_value())]
+            smallvec![
+                // Bearer token was moved to its own regest and type out of the passwords, but we
+                // still keep it here for backwards compatibility.
```
I think this is correct:

```rs
(v, &*PASSWORD_KEY_REGEX, ReplaceBehavior::replace_value()),
```

Is what Cursor wants you to use (`v` instead of `kv`), which seems reasonable.

olksdr replied:
You mean for Bearer ?

Dav1dde replied:
ah oops, copied the wrong one, yeah for Bearer

olksdr committed:
Scrub bearer in values only

olksdr committed:
Merge remote-tracking branch 'origin/master' into fix/pii-logentry-formatted

Dav1dde commented on CHANGELOG.md:
```diff
@@ -5,6 +5,7 @@
 **Features**:
 
 - Always emit a span usage metric, independent of span feature flags. ([#4976](https://github.com/getsentry/relay/pull/4976))
+- Avoid replacing the entire value of `logentry.formatted` during PII scrubbing. ([#4985](https://github.com/getsentry/relay/pull/4985))
```
`Avoid replacing` implies we have been replacing it already, but before it wasn't scrubbed at all.

olksdr replied:
yeah, thanks to @loewenheim  👼  
I will try to change this a bit more

Dav1dde commented on relay-pii/src/processor.rs:
```diff
@@ -1182,6 +1181,84 @@ mod tests {
+
+        let mut processor = PiiProcessor::new(config.compiled());
+        process_value(&mut event, &mut processor, ProcessingState::root()).unwrap();
+        assert_annotated_snapshot!(event);
```
I am a fan of having these snapshots inlined, makes it easier to review, but also nbd if you prefer this.

olksdr replied:
I can look into it, if that's easier to understand 👍

Dav1dde commented on relay-pii/src/convert.rs:
```diff
@@ -307,7 +331,7 @@ THd+9FBxiHLGXNKhG/FRSyREXEt+NyYIf/0cyByc9tNksat794ddUqnLOg0vwSkv
         });
 
-        insta::assert_json_snapshot!(pii_config, @r###"
+        insta::assert_json_snapshot!(pii_config, @r#"
```
I think `###` is the 'new' syntax insta wants and `cargo insta --force-update-snapshots` would change it back.

olksdr replied:
I always thought it's another way around, but I will run it and see how changes: 👍

olksdr replied:
hm, it seems like I do not have this option `--force-update-snapshots` in my local installation

olksdr replied:
ok, found 😄

olksdr replied:
So it seems like i was right  and running latest version of insta with `cargo insta test --force-update-snapshots --all-features --workspace`
```
cargo-insta 1.43.1
```
changes all the tests to single `#` or remove it all together when needed. 

@Dav1dde  should be do it in separate PR to match the latest insta and , make that change separate from this  PR ?

Dav1dde replied:
That's strange, when I ran it, it did the opposite: https://github.com/getsentry/relay/pull/4908/files

Dav1dde replied:
Okay, I am on an older version and they changed behavior in 1.41 -_-

yeah let's keep the insta stuff separate 👍

Dav1dde commented on relay-pii/src/regexes.rs:
```diff
@@ -324,9 +332,11 @@ regex!(
     "
 );
 
+regex!(BEARER_TOKEN_REGEX, r"(?i)\b(Bearer\s+)([^\s]+)");
```
The `\b` is there to make sure `foobearer` doesn't match?

olksdr replied:
yeah, basically to define the word boundary

Dav1dde commented on tests/integration/test_pii.py:
```diff
@@ -35,3 +35,246 @@ def test_scrub_span_sentry_tags_advanced_rules(mini_sentry, relay):
     assert event["spans"][0]["sentry_tags"]["user.geo.subregion"] == "**"
+
+
+def test_logentry_formatted_smart_scrubbing_email(mini_sentry, relay):
```
These tests are very similar maybe worth consolidating into a single parameterized test.

olksdr replied:
makes sense, thanks! Will do

Dav1dde commented on relay-pii/src/convert.rs:
```diff
@@ -71,6 +71,23 @@ pub fn to_pii_config(
             vec!["@anything:filter".to_owned()],
         );
+
+        let logentry_selector: SelectorSpec = SelectorSpec::Path(vec![
+            SelectorPathItem::Type(ValueType::LogEntry),
+            SelectorPathItem::Key("formatted".to_owned()),
+        ]);
```
I think I would wanna apply the same rule for an ourlog body.

olksdr committed:
Move py tests to parametrized

olksdr committed:
review comments: inlined snapshots and changelog

olksdr committed:
Merge remote-tracking branch 'origin/master' into fix/pii-logentry-formatted
